haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs

```diff
@@ -380,7 +380,19 @@ type family BaseMessage (t :: MethodType) :: Method p t -> Type where
   BaseMessage Request = RequestMessage
   BaseMessage Notification = NotificationMessage
 
-type HandlerFunc a = Either ResponseError a -> IO ()
+-- | The type of a handler that handles requests and notifications coming in
+-- from the server or client
+type family Handler m :: Type where
+  Handler (m :: Method p t) = BaseHandler m (IO ())
+
+-- | Version of 'Handler' that can be used to construct arbitrary functions
+-- taking in the required handler arguments
+type family BaseHandler (m :: Method p t) a :: Type where
```

Comment:
Handler and BaseHandler seem to be almost the same thing. Do we need both?

Reply:
I'd like to roll them into one but I'm not sure how to pattern match on both the `m :: Method p t` and the `t :: MethodType` simultaneously in one family, any ideas?

Reply:
but you aren't doing this in `Handler` or `BaseHandler`?

Reply:
Huh no idea what I was doing earlier but this seems to work now, we can probably remove `BaseMessage` as well then
```haskell
type family BaseHandler (m :: Method p t) (a :: Type) :: Type where
  BaseHandler (m :: Method p Request)      a = RequestMessage m -> (Either ResponseError (ResponseParams m) -> IO ()) -> a
  BaseHandler (m :: Method p Notification) a = NotificationMessage m -> a
```